### PR TITLE
Remove unused header files

### PR DESCRIPTION
**Pull request by xingwozhonghua126**

### CODE REVIEW

File: src/conftests/sdl2_mixer.cpp [hunk 1]
```diff
@@ -13,7 +13,6 @@
 */
 
 #include <SDL2/SDL.h>
-#include <SDL2/SDL_mixer.h>
```
Comment (stevecotton) [inline src/conftests/sdl2_mixer.cpp:16]: This one's necessary - conftests are checking which headers are installed, and on Debian systems `libsdl2-mixer-dev` is a separate package.

**Comment by stevecotton** [comment]
It's looking good. The `cstdlib` includes appear to have been added for an old version of Xcode, the others I can see why they're unused.

Still waiting for the CI to finish, and will leave open for comments for another 24 hours before merging.

**Comment by CelticMinstrel** [comment]
Honestly this feels slightly dangerous, but… as long as the contents of those includes _aren't_ directly used in the relevant files, I guess it should be fine.

**Comment by xingwozhonghua126** [comment]
I could have cleaned up a lot of useless header files, but you said in IRC that it is not recommended to modify it like this, so I don't dare to continue to clean up, so I will try to find other problems and modify it

**Comment by Wedge009** [comment]
Why dangerous? If the CI checks pass (and they have) shouldn't that mean at least building across the supported platforms is successful without the headers?

**Comment by stevecotton** [comment]
The CI probably isn't the full set of platforms that we support - if any dev or packager said that their platform needed a particular `#include` to be kept, we'd probably re-add it; technically that's Pentarctagon's decision to make, but I expect we would. We definitely have added headers for upcoming versions of Boost, when Boost has itself tidied up its header includes, and thus an indirect include no longer works on the development version of a Linux distro.

Not sure what to recommend for @xingwozhonghua126 to look at - help is welcome, but the next month-and-a-half until 1.18.0 releases should be more focused on bugfixing.

**Comment by Pentarctagon** [comment]
We have on occasion made changes to support FreeBSD for example, which isn't in the CI. As always it depends on the situation though.

The best thing for right now though would be to check the [Good First Issue](https://github.com/wesnoth/wesnoth/issues?q=is%3Aopen+is%3Aissue+label%3A%22Good+first+issue%22) tag and then pick one that's a bug and which doesn't require a translatable string change to fix.

**Comment by CelticMinstrel** [comment]
> Why dangerous? If the CI checks pass (and they have) shouldn't that mean at least building across the supported platforms is successful without the headers?

This is just one place where I'm not sure if the CI environments can represent every possible supported environment. If you've taken the time to check the entire file and verify that things defined in those headers are not directly referenced anywhere in the file, then it's probably fine.

It's not absolutely the end of the world if this breaks someone's build, but it would be annoying for them to have to go and put the header back in just so they can build.

> The CI probably isn't the full set of platforms that we support - if any dev or packager said that their platform needed a particular `#include` to be kept, we'd probably re-add it; technically that's Pentarctagon's decision to make, but I expect we would. We definitely have added headers for upcoming versions of Boost, when Boost has itself tidied up its header includes, and thus an indirect include no longer works on the development version of a Linux distro.

I think such issues stem from undocumented dependencies between standard headers – for example, `<vector>` might implicitly include `<utility>` on some platforms but not on others, or `<cstdlib>` might implicitly include `<cstring>` on some platforms but not on others. Therefore, if you scan through the entire file for everything _officially_ defined by the removed header and find no matches, it shouldn't break any standard-conforming platform. This is obviously a lot more work than just hitting Build, however.

**Comment by stevecotton** [comment]
While removing includes of system headers can cause cross-platform problems, that shouldn't be the case for removal of Wesnoth's own headers, and the latter would be useful for developer time because it reduces the number of .cpp files that need to be rebuilt when editing a .hpp file.

@xingwozhonghua126, I'm not sure how much cleanup is possible, but please would you use your knowledge of this tool to see if it looks like there's a reasonable improvement available?

Note: there's some longstanding technical debt that's so big that even the regular contributors would only work on it piecewise, so stop quickly if a particular header looks like too much effort. For example, the [number of non-UI classes that include display_context.hpp](https://devdocs.wesnoth.org/display__context_8hpp__dep__incl_org.svg), which despite sounding like a UI class is really [a bunch of sort-of singletons that tend to be accessed via display::get_singleton()](https://devdocs.wesnoth.org/classdisplay__context.html#details).

**Comment by CelticMinstrel** [comment]
> that shouldn't be the case for removal of Wesnoth's own headers

It's mostly correct that there's no issue with removing includes of headers defined in the project. It's not impossible for such a thing to cause errors (because that header included a system header that the file was relying on), but if that _does_ occur, the correct action would be to put back just that system header, so removing the original include is still fine.

**Comment by xingwozhonghua126** [comment]
I scanned it out with IWYU, and some header file modifications suggested that I make a preliminary evaluation according to the header file to be deleted, and then try to compile after deletion, if it can be compiled normally, I think this header file is invalid. to delete. There are a lot of scans, can I continue to submit them in the future? If I can, I can commit directory by directory.

**Comment by CelticMinstrel** [comment]
I would assume that this IWYU tool is written with the intent of finding references to things that are officially defined in each header, rather than things that are _actually_ defined in each header; either that, or finding things that are _directly_ defined in the header while excluding anything pulled in from transitive standard includes (but including transitive _non-standard_ includes).

So if it tells you a header can definitely be deleted, then it's _probably_ fine to do so.

**Comment by stevecotton** [comment]
> I would assume that this IWYU tool is written with the intent of finding references to things that are officially defined in each header, rather than things that are _actually_ defined in each header

The opposite - it's just working out where things are defined, and relies on comments or a separate stand-alone list to know which headers are meant to be private implementation details. So it sounds like it needs someone experienced with it to understand which suggestions are worth pursuing.

https://github.com/include-what-you-use/include-what-you-use/blob/master/README.md

As @xingwozhonghua126 already knows and uses the tool, it seems practical to see what the results are.

**Comment by CelticMinstrel** [comment]
> a separate stand-alone list

Surely this already exists for at least the standard libraries, though?

**Pull request merged**